Source/SwiftLintFramework/Rules/Idiomatic/ReduceIntoRule.swift

```diff
@@ -0,0 +1,35 @@
+import SourceKittenFramework
+
+public struct ReduceIntoRule: ConfigurationProviderRule, OptInRule, AutomaticTestableRule {
+    public var configuration = SeverityConfiguration(.warning)
+
+    public init() {}
+
+    public static var description = RuleDescription(
+        identifier: "reduce_into",
+        name: "Reduce Into",
+        description: "Prefer `reduce(into:_:)` over `reduce(_:_:)`",
+        kind: .performance,
+        minSwiftVersion: .three,
+        nonTriggeringExamples: [
+            "let matches = lines.reduce(into: []) { matches, line in",
+            "zip(group, group.dropFirst()).reduce(into: []) { result, pair in",
+            "values.reduce(into: \"\") { $0.append(\"\\($1)\") }"
+        ],
+        triggeringExamples: [
+            "let matches = lines.↓reduce([]) { matches, line in",
+            "zip(group, group.dropFirst()).↓reduce([]) { result, pair in",
+            "values.↓reduce(\"\") { $0 + \"\\($1)\" }"
+        ]
+    )
+
+    public func validate(file: File) -> [StyleViolation] {
+        let pattern = "reduce\\((?!into:)"
```

Comment:
what do you think about implementing this as an `ASTRule` instead?

You can check the method name and check that the first parameter is a COW structure (a collection or string literal).

Reply:
Oh interesting. I'll look into this.

Reply:
Let me know if you need help or have any questions!

Reply:
Just updated. I'm sure my implementation isn't exactly what you meant, but I'll need a bit more direction on how to proceed. I'm happy to keep tweaking it if you'll clarify a bit more, though.

Reply:
I think this is a good start! 

I'd also check if the first parameter is one of the following:

- array literal: `kind == source.lang.swift.expr.array` (we have this as an enum in `SwiftExpressionKind`)
- dictionary literal: `kind == source.lang.swift.expr.dictionary`
- string literal: this one is trickier because you'll need to use the `syntaxMap` and look for a string syntax type inside the first argument range
- an array or dictionary init, e.g `[Int]()`, `Array<Int>()`, `[Int: String]()`, `Dictionary<Int, String>()`

You can check https://github.com/realm/SwiftLint/pull/2657/files for some real-world examples.

Reply:
Thanks for the feedback! I'll keep iterating.